luatest/replica_set.lua

```diff
@@ -0,0 +1,231 @@
+--
+-- @classmod luatest.replica_set
+
+local checks = require('checks')
+local log = require('log')
+
+local helpers = require('luatest.helpers')
+local Server = require('luatest.server')
+
+local ReplicaSet = {}
+
+function ReplicaSet:inherit(object)
+    setmetatable(object, self)
+    self.__index = self
+    return object
+end
+
+--- Build a replica set object.
+--
+-- @tab[opt] object
+-- @tab[opt] object.servers List of server configurations to build server
+--   objects from and add them to the new replica set. See an example below.
+-- @return table
+-- @see luatest.server:new
+-- @usage
+--   local ReplicaSet = require('luatest.replica_set')
+--   local Server = require('luatest.server')
+--   local box_cfg = {
+--       replication_timeout = 0.1,
+--       replication_connect_timeout = 10,
+--       replication_sync_lag = 0.01,
+--       replication_connect_quorum = 3,
+--       replication = {
+--           Server.build_listen_uri('replica1'),
+--           Server.build_listen_uri('replica2'),
+--           Server.build_listen_uri('replica3'),
+--       },
+--   }
+--   local replica_set = ReplicaSet:new({
+--       servers = {
+--           {alias = 'replica1', box_cfg = box_cfg},
+--           {alias = 'replica2', box_cfg = box_cfg},
+--           {alias = 'replica3', box_cfg = box_cfg},
+--       }
+--   })
+--   replica_set:start()
+--   replica_set:wait_for_fullmesh()
+function ReplicaSet:new(object)
+    if not object then object = {} end
+    self:inherit(object)
+    object:initialize()
+    return object
+end
+
+-- Initialize the replica set object.
+function ReplicaSet:initialize()
+    self._server = Server
+    if self.servers then
+        local configs = table.deepcopy(self.servers)
+        self.servers = {}
+        for _, config in ipairs(configs) do
+            self:build_and_add_server(config)
+        end
+    else
+        self.servers = {}
+    end
+end
+
+--- Build a server object for the replica set.
+--
+-- @tab[opt] config Configuration for the new server.
+-- @return table
+-- @see luatest.server:new
+function ReplicaSet:build_server(config)
+    checks('table', self._server.constructor_checks)
+    if config then config = table.deepcopy(config) end
+    return self._server:new(config)
+end
+
+--- Add the server object to the replica set.
+--
+-- @tab server Server object to be added to the replica set.
+function ReplicaSet:add_server(server)
+    checks('table', 'table')
+    if self:get_server(server.alias) then
+        error(('Server with alias "%s" already exists in replica set')
+            :format(server.alias))
+    end
+    table.insert(self.servers, server)
+end
+
+--- Build a server object and add it to the replica set.
+--
+-- @tab[opt] config Configuration for the new server.
+-- @return table
+-- @see luatest.server:new
+function ReplicaSet:build_and_add_server(config)
+    checks('table', self._server.constructor_checks)
```

Comment:
this check is a part of the `build_server(object)` method. Why make it just before calling that method?

Reply:
Okay, I thought it was some kind of assert. Is it something else?

Reply:
understood, thanks